File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.js
function solution(sequence) {
    let maxSum = 0;

    let curr1 = 0; // [1, -1, 1, -1 ...]
    let curr2 = 0; // [-1, 1, -1, 1 ...]

    for (let i = 0; i < sequence.length; i++) {
        const sign1 = (i % 2 === 0) ? 1 : -1;
        const sign2 = -sign1;

        const v1 = sequence[i] * sign1;
        const v2 = sequence[i] * sign2;

        curr1 = Math.max(v1, curr1 + v1);
        curr2 = Math.max(v2, curr2 + v2);

        maxSum = Math.max(maxSum, curr1, curr2);
    }

    return maxSum;
}

File: 프로그래머스/3/161988. 연속 펄스 부분 수열의 합/연속 펄스 부분 수열의 합.java
class Solution {
    public long solution(int[] sequence) {
        long maxSum = 0;

        long curr1 = 0; // [1, -1, 1, -1 ...]
        long curr2 = 0; // [-1, 1, -1, 1 ...]

        for (int i = 0; i < sequence.length; i++) {
            long sign1 = (i % 2 == 0) ? 1 : -1;
            long sign2 = -sign1;

            long v1 = sequence[i] * sign1;
            long v2 = sequence[i] * sign2;

            curr1 = Math.max(v1, curr1 + v1);
            curr2 = Math.max(v2, curr2 + v2);

            maxSum = Math.max(maxSum, Math.max(curr1, curr2));
        }

        return maxSum;
    }
}
